logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueReader.java

```diff
@@ -271,6 +321,44 @@ public void setCurrentReaderAndPosition(Path segmentPath, long position) throws
                 .ifPresent(reader -> currentReader = reader);
     }
 
+    private void removeSegmentsBefore(Path validSegment) throws IOException {
+        final Comparator<Path> fileTimeAndName = ((Comparator<Path>) this::byFileFimestamp)
+                .thenComparingInt(DeadLetterQueueUtils::extractSegmentId);
+
+        Files.list(queuePath)
+                .filter(p -> fileTimeAndName.compare(p, validSegment) < 0)
+                .forEach(this::trashSegment);
+    }
+
+    private int byFileFimestamp(Path p1, Path p2) {
+        FileTime timestamp1;
+        // if one of the getLastModifiedTime raise an error, consider them equals
+        // and fallback to the other comparator
+        try {
+            timestamp1 = Files.getLastModifiedTime(p1);
+        } catch (IOException ex) {
+            logger.warn("Error reading file's timestamp for {}", p1, ex);
+            return 0;
+        }
+
+        FileTime timestamp2;
+        try {
+            timestamp2 = Files.getLastModifiedTime(p2);
+        } catch (IOException ex) {
+            logger.warn("Error reading file's timestamp for {}", p2, ex);
+            return 0;
+        }
+        return timestamp1.compareTo(timestamp2);
+    }
+
+    private void trashSegment(Path segment) {
```

Comment:
Nit: Maybe deleteSegment would better indicate intent of method
```suggestion
    private void deleteSegment(Path segment) {
```